pkg/controller/master/node/promote_controller.go

```diff
@@ -0,0 +1,511 @@
+	promoteNodeNumber := specMasterNumber - masterNumber
+	if promoteNodeNumber <= 0 {
+		return nil, nil
+	}
+
+	// sort by creation time from largest to smallest to find the youngest node that can be promoted
+	promoteNodeList := make([]*corev1.Node, 0, promoteNodeNumber)
+	var sortByAgeNodeList SortByAgeNodeList = canBePromoteNodeList
+	sort.Sort(sortByAgeNodeList)
+	for _, node := range sortByAgeNodeList {
+		if promoteNodeNumber > 0 {
+			promoteNodeList = append(promoteNodeList, node)
+			promoteNodeNumber--
+		}
+	}
+
+	return promoteNodeList, nil
+}
+
+// getSpecMasterNumber get spec master number by all node number
+func (h *PromoteHandler) getSpecMasterNumber(nodeNumber int) int {
+	if nodeNumber < 3 {
+		return 1
+	}
+	return defaultSpecMasterNumber
+}
+
+// isHealthyNode determine whether it's an healthy node
+func (h *PromoteHandler) isHealthyNode(node *corev1.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Type == corev1.NodeReady && c.Status != corev1.ConditionTrue {
+			// skip unready nodes
+			return false
+		}
+
+		if c.Type != corev1.NodeReady && c.Status == corev1.ConditionTrue {
+			// skip node with conditions like nodeMemoryPressure, nodeDiskPressure, nodePIDPressure
+			// and nodeNetworkUnavailable equal to true
+			return false
+		}
+	}
+	return true
+}
+
+// isHarvesterNode determine whether it's an Harvester node based on the node's label
+func (h *PromoteHandler) isHarvesterNode(node *corev1.Node) bool {
+	_, ok := node.Labels[HarvesterVersionLabelKey]
+	return ok
+}
+
+// shouldBeMaster determine whether it should be an master node
+func (h *PromoteHandler) shouldBeMaster(node *corev1.Node) bool {
+	return h.isMasterRole(node) || h.hasBeenPromoted(node)
+}
+
+// isMasterRole determine whether it's an master node based on the node's label
+func (h *PromoteHandler) isMasterRole(node *corev1.Node) bool {
+	if value, ok := node.Labels[KubeMasterNodeLabelKey]; ok {
+		return value == "true"
+	}
+
+	return false
+}
+
+func (h *PromoteHandler) hasBeenPromoted(node *corev1.Node) bool {
+	return h.hasPromoteJob(node) || h.hasPromoteStatus(node)
+}
+
+func (h *PromoteHandler) hasPromoteJob(node *corev1.Node) bool {
+	jobName := buildPromoteJobName(node.Name)
+	_, err := h.jobCache.Get(promoteJobNamespaceName, jobName)
+	return err == nil
+}
+
+func (h *PromoteHandler) hasPromoteStatus(node *corev1.Node) bool {
+	_, ok := node.Annotations[HarvesterPromoteStatusAnnotationKey]
+	return ok
+}
+
+func (h *PromoteHandler) isPromoteStatusIn(node *corev1.Node, statuses ...string) bool {
+	status, ok := node.Annotations[HarvesterPromoteStatusAnnotationKey]
+	if !ok {
+		return false
+	}
+
+	for _, s := range statuses {
+		if status == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+func checkError(action string, err error) error {
+	if err != nil {
+		logrus.Errorf("%s failed, %s", action, err)
+		return err
+	}
+	logrus.Infof("%s", action)
+	return nil
+}
```

Comment:
Some cons of this pattern:
1. We usually return errors to the caller and don't log the error message directly(The caller handles it). Otherwise it may result in duplicate error logs. you can return a wrapped error if you want to provide more info.
2. every call for this generates an info level log, but looks like some implementation details users may not need to know?
3. It is called checkError but seems to be a logging util function.

In short, I would suggest we do the error check and logging directly instead of wrapping them here